### ISSUE TITLE
Nozzle Tip Runout Calibration

### ISSUE THREAD

**Issue by vonnieda** [openpnp/openpnp#235]
I'm noticing a lot of runout on my nozzles and others are too. This severely degrades the performance bottom vision ( #104 ) along with general placement. We need to be able to compensate for runout by checking the runout with the up looking camera and interpolating X / Y movements.

Code for this was submitted a long time ago in #57 but it was not merged due to the complexities of the change. That code can be used as a reference, along with the good circle finding we now have throughout the system.

My initial thought on this is:
- Whenever a Nozzle is commanded to move, if it has not been calibrated it should calibrate first.
- Nozzle tip change should reset the calibration status.
- It should be possible to enable / disable calibration through the UI.
- I think most of the changes can be done by branching off of ReferenceNozzle.moveTo.

**Comment by cri-s** [comment]
>Whenever a Nozzle is commanded to move, if it has not been calibrated it should calibrate first.
It should be calibrated using component, checking nozzle only have some issues.
>Nozzle tip change should reset the calibration status
This is not true, if automaitc changer is used, as in this case it is unload and load using rotation 0.
It's thinkable to use red dots/color, and filter that out in order to check if nozzle position is ok or not,
in the event nozzle was removed manually.

**Comment by vonnieda** [comment]
This code is now in and working, but it's not yet documented at all. Leaving this ticket open until that's done.

**Comment by vonnieda** [comment]
Note: Previous comment said this is working, but that may not be the case. More testing is needed, along with documentation and any fixes required.

**Comment by vonnieda** [comment]
See comments from @cri-s on #337 for potential issues in the current implementation.

**Comment by vonnieda** [comment]
Also see comments on #336 for additional issues related to angle and rotation.

**Comment by vonnieda** [comment]
This feature has temporarily been removed from the configuration UI since it is not working correctly and was causing confusion. See https://github.com/openpnp/openpnp/commit/5f586166bb74a81b11a78c44a15909ebc81b2bf2

**Comment by vonnieda** [comment]
@markmaker The issue was related to how OpenPnP was reading coordinates from the nozzles when the compensation was applied with regards to bottom vision. Basically, there is a stack of transforms that were being applied and it seems like some were either incorrect or backwards. The comments were not particularly clear on the error, so I think it's best that someone interested in this feature dig it out and do some testing.

**Comment by ldpgh** [comment]
ReferenceNozzleTip.java (develop-branac @ line 456 ... it seems to be divison-by-0
1. develop-branch
2. enabled calibration panel in ReferenceNozzleTipConfigurationWizard.java @ line 382
3. ReferenceNozzleTip @ line 420 ... lowered 360 to 3

![image](https://user-images.githubusercontent.com/13517069/46910309-c124b680-cf42-11e8-886c-9efce10be8a3.png)

**Comment by ldpgh** [comment]
Connect Machine
1. ReferenceNozzleTip-NT2_2MM67 is still loaded to the Nozzle-N1
MachineSetup->Heads->ReferenceHead-H1->ReferenceNozzle-N1->NozzleTips->ReferenceNozzleTip-NT2_2MM67
1. Calibrate Enabled
2. Apply button
3. Press Calibrate button 
MachineSetup->Cameras->OpenPnpCaptureCamera-OpenPnpCaptureCameraUp / Location
1. Position Camera over center of location
2. Position Tool over center of location -> NaN for X and Y ... see log below


```
2018-10-14 00:03:24 Main DEBUG: Bienvenue, Bienvenido, Willkommen, Hello, Namaskar, Welkom to OpenPnP version INTERNAL BUILD.
2018-10-14 00:03:24 Scripting TRACE: Scripting.on Startup
2018-10-14 00:03:31 ReferenceMachine DEBUG: setEnabled(true)
2018-10-14 00:03:31 GcodeDriver DEBUG: sendCommand(null, 250)...
2018-10-14 00:03:31 GcodeDriver DEBUG: sendCommand(serial://COM220 null, 250) => []
2018-10-14 00:03:31 GcodeDriver DEBUG: sendCommand(connect, 5000)...
2018-10-14 00:03:31 GcodeDriver TRACE: [serial://COM220] >> connect
2018-10-14 00:03:31 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:03:31 GcodeDriver DEBUG: sendCommand(serial://COM220 connect, 5000) => [ok]
2018-10-14 00:03:31 GcodeDriver DEBUG: sendCommand(enable, 5000)...
2018-10-14 00:03:31 GcodeDriver TRACE: [serial://COM220] >> enable
2018-10-14 00:03:35 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:03:35 GcodeDriver DEBUG: sendCommand(serial://COM220 enable, 5000) => [ok]
2018-10-14 00:03:35 ReferenceHead DEBUG: H1.home()
2018-10-14 00:03:35 GcodeDriver DEBUG: sendCommand(home, -1)...
2018-10-14 00:03:35 GcodeDriver TRACE: [serial://COM220] >> home
2018-10-14 00:03:39 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:03:39 GcodeDriver DEBUG: sendCommand(serial://COM220 home, -1) => [ok]
2018-10-14 00:03:40 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:41 ReferenceHead DEBUG: H1.moveToSafeZ(0.86)
2018-10-14 00:04:41 ReferenceNozzle DEBUG: N1.moveToSafeZ(0.86)
2018-10-14 00:04:41 GcodeDriver DEBUG: sendCommand(moveto   Z42.0000  F860, 5000)...
2018-10-14 00:04:41 GcodeDriver TRACE: [serial://COM220] >> moveto   Z42.0000  F860
2018-10-14 00:04:41 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:41 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto   Z42.0000  F860, 5000) => [ok, ok]
2018-10-14 00:04:41 ReferenceCamera DEBUG: OpenPnpCaptureCameraDown.moveToSafeZ(0.86)
2018-10-14 00:04:41 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_01.moveToSafeZ(0.86)
2018-10-14 00:04:41 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_02.moveToSafeZ(0.86)
2018-10-14 00:04:41 ReferenceNozzle DEBUG: N1.moveTo((53.373000, 175.606000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:41 GcodeDriver DEBUG: sendCommand(moveto X84.2620 Y174.7980   F860, 5000)...
2018-10-14 00:04:41 GcodeDriver TRACE: [serial://COM220] >> moveto X84.2620 Y174.7980   F860
2018-10-14 00:04:44 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:44 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto X84.2620 Y174.7980   F860, 5000) => [ok]
2018-10-14 00:04:44 ReferenceNozzle DEBUG: N1.moveTo((53.373000, 175.606000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:44 Scripting TRACE: Scripting.on Camera.BeforeSettle
2018-10-14 00:04:45 Scripting TRACE: Scripting.on Camera.BeforeCapture
2018-10-14 00:04:45 Scripting TRACE: Scripting.on Camera.AfterCapture
2018-10-14 00:04:45 ReferenceNozzle DEBUG: N1.moveTo((53.310000, 175.291000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:45 GcodeDriver DEBUG: sendCommand(moveto X84.1990 Y174.4830   F860, 5000)...
2018-10-14 00:04:45 GcodeDriver TRACE: [serial://COM220] >> moveto X84.1990 Y174.4830   F860
2018-10-14 00:04:45 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:45 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto X84.1990 Y174.4830   F860, 5000) => [ok]
2018-10-14 00:04:45 Scripting TRACE: Scripting.on Camera.BeforeSettle
2018-10-14 00:04:45 Scripting TRACE: Scripting.on Camera.BeforeCapture
2018-10-14 00:04:45 Scripting TRACE: Scripting.on Camera.AfterCapture
2018-10-14 00:04:45 ReferenceNozzle DEBUG: N1.moveTo((53.319000, 175.300000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:45 GcodeDriver DEBUG: sendCommand(moveto X84.2080 Y174.4920   F860, 5000)...
2018-10-14 00:04:45 GcodeDriver TRACE: [serial://COM220] >> moveto X84.2080 Y174.4920   F860
2018-10-14 00:04:46 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:46 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto X84.2080 Y174.4920   F860, 5000) => [ok]
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.BeforeSettle
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.BeforeCapture
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.AfterCapture
2018-10-14 00:04:46 ReferenceNozzle DEBUG: N1.moveTo((53.292000, 175.273000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:46 GcodeDriver DEBUG: sendCommand(moveto X84.1810 Y174.4650   F860, 5000)...
2018-10-14 00:04:46 GcodeDriver TRACE: [serial://COM220] >> moveto X84.1810 Y174.4650   F860
2018-10-14 00:04:46 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:04:46 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto X84.1810 Y174.4650   F860, 5000) => [ok]
2018-10-14 00:04:46 ReferenceNozzle DEBUG: N1.moveTo((53.292000, 175.273000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.BeforeSettle
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.BeforeCapture
2018-10-14 00:04:46 Scripting TRACE: Scripting.on Camera.AfterCapture
2018-10-14 00:04:46 ReferenceNozzle DEBUG: N1.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceHead DEBUG: H1.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceNozzle DEBUG: N1.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceCamera DEBUG: OpenPnpCaptureCameraDown.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_01.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_02.moveToSafeZ(0.86)
2018-10-14 00:05:47 ReferenceCamera DEBUG: moveTo((53.373000, 175.606000, NaN, 0.000000 mm), 0.86)
2018-10-14 00:05:47 GcodeDriver DEBUG: sendCommand(moveto X53.3730 Y175.6060   F860, 5000)...
2018-10-14 00:05:47 GcodeDriver TRACE: [serial://COM220] >> moveto X53.3730 Y175.6060   F860
2018-10-14 00:05:48 GcodeDriver TRACE: [serial://COM220] << ok
2018-10-14 00:05:48 GcodeDriver DEBUG: sendCommand(serial://COM220 moveto X53.3730 Y175.6060   F860, 5000) => [ok]
2018-10-14 00:05:48 ReferenceCamera DEBUG: moveTo((53.373000, 175.606000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:05:51 ReferenceHead DEBUG: H1.moveToSafeZ(0.86)
2018-10-14 00:05:51 ReferenceNozzle DEBUG: N1.moveToSafeZ(0.86)
2018-10-14 00:05:51 ReferenceCamera DEBUG: OpenPnpCaptureCameraDown.moveToSafeZ(0.86)
2018-10-14 00:05:51 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_01.moveToSafeZ(0.86)
2018-10-14 00:05:51 ReferenceActuator DEBUG: ZZZ_ReferenceActuator_Dummy_02.moveToSafeZ(0.86)
2018-10-14 00:05:51 ReferenceNozzle DEBUG: N1.moveTo((53.373000, 175.606000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:05:51 ReferenceNozzle DEBUG: N1.moveTo((NaN, NaN, 42.000000, 0.000000 mm), 0.86) (corrected)
2018-10-14 00:05:51 ReferenceNozzle DEBUG: N1.moveTo((53.373000, 175.606000, 42.000000, 0.000000 mm), 0.86)
2018-10-14 00:05:51 ReferenceNozzle DEBUG: N1.moveTo((NaN, NaN, 42.000000, 0.000000 mm), 0.86) (corrected)
```

**Comment by ldpgh** [comment]
ReferenceNozzleTip.java @ line 456 . . . overwritten by `double ratio=1.0;` to circumvent divison-by-0.

Connect Machine
1. ReferenceNozzleTip-NT2_2MM67 is still loaded to the Nozzle-N1

MachineSetup->Heads->ReferenceHead-H1->ReferenceNozzle-N1->NozzleTips->ReferenceNozzleTip NT2_2MM67
1. Calibrate Enabled
2. Apply button
3. Press Calibrate button

MachineSetup->Cameras->OpenPnpCaptureCamera-OpenPnpCaptureCameraUp / Location
1. Position Camera over center of location
2.1. Position Tool over center of location -> The original runout is back <<<<<<<<<<<<<<<<<
2.2. Expectated is to have the calibration incorporated. Maybe this assumption is not valid.

**Comment by ldpgh** [comment]
It seems the basic offset between nozzle-tip and nozzle (at 0 degree / with `camera.getLocation()` as reference ) is not kept.

In ReferenceNozzleTip.java:405...415 the nozzle-tip is centered and this location is stored in `Location startLocation=location;`. The mentioned basic offset would be `location-camera.getLocation()`.

Though `startLocation` is used as reference to derive the angle-dependent offset (runout, lines 419...429), it seems the basic offset from the nozzle-tip to the nozzle is not kept outside the method `calibrate()`.

**Comment by ldpgh** [comment]
To illustrate the previous comment ... some screenshots.
(1) Power Button
(2) MachineSetup->Cameras->OpenPnpCaptureCamera / CameraSpecific->Position tool over center of location ... with installed nozzle-tip has calibration enabled.

OpenPnp started ... still disconnected
![calib_start_1](https://user-images.githubusercontent.com/13517069/46918094-edd1e000-cfcd-11e8-8c5f-9a1a6bea66d4.png)
Power Button pressed ... it connects the machine and issues a `home()` command

End of calibrate() methode ... right after last log-message
New postion looks nice :-)
![calib_endofcalibmethod_2](https://user-images.githubusercontent.com/13517069/46918144-7b153480-cfce-11e8-9095-c6d7a83ba75d.png)

calibrate() methode returned
Before line 243 @  ReferenceNozzle.java:moveTo()
![calib_calibmethodreturned_3](https://user-images.githubusercontent.com/13517069/46918145-823c4280-cfce-11e8-9565-905bec107766.png)

After calling line 243 @ ReferenceNozzle.java:moveTo()
![calib_finalmoveto_4](https://user-images.githubusercontent.com/13517069/46918102-03dfa080-cfce-11e8-9d49-e2d930cc4175.png)

Pogramm has returned to the GUI
![calib_returnedtogui_5](https://user-images.githubusercontent.com/13517069/46918106-0b06ae80-cfce-11e8-8420-18c3fbced1d9.png)

**Comment by aneox** [comment]
I have problem with bottom light. 
When camera is using for part alignment, bottom lights turns on in Vision.PartAlignment.Before.js and turns off in Vision.PartAlignment.After.js. Works good.

When camera is using for Nozzle Tip Runout Calibration bottom lights turns on in Camera.BeforeCapture.js and turns off in Camera.AfterCapture.js. But in this case, for some reason, calibration starts before lights turns on and calibration fails. If turn on lights manually, all works good.

**Comment by vonnieda** [comment]
This is a summary of the comments that were deleted from #337 and #336.

CC: @ldpgh, @markmaker 

Previous brief summary: https://github.com/openpnp/openpnp/issues/235#issuecomment-423006170

These issues have not been verified. They were posted and then due to this being a new feature I removed the feature because it sounded like there were issues. It's possible that these are not issues at all, but as I mentioned in the comment above, we should investigate and validate that the feature works well.

* There is a potential divide by zero during the final angle increment.
* There was a comment that the nozzle moves in X and Y during calibration which may introduce backlash errors. This is because we check and center multiple times to get the best centerpoint. This is required when the camera is not calibrated, but maybe it would be better if we avoid it. Worth testing both ways.
* Concern that the DROs are incorrect after calibration, or that the "capture coordinates" buttons do not get valid coordinates. Should be easy to check by introducing a large calibration offset.
* A comment, "Nozzle position center is not used and as result calibration has offsets.". I'm not sure what this meant and never got an explanation.

So, given the concerns above, what really needs to happen is someone take ownership of this issue / feature, finish developing it and really test it. Don't consider the current code to be ideal - perhaps it can be much better. Maybe the algorithm can be improved. Consider the current code to be a starting point. My initial implementation of it was just a first attempt, so maybe we can do much better.

**Comment by netzmark** [comment]
I have unlocked the calibration panel in the ReferenceNozzleTipConfigurationWizard to play with the feature a bit and consider why can't see the fields X,Y,Z,Rotation in the Calibration section?
Was this added by ldpgh or markmaker while their testings?

**Comment by netzmark** [comment]
I can't see x,y,z,c fields (windows) not values inside them.
Should I see an explanation around the comment above?

**Comment by ldpgh** [comment]
Are you running #openpnp IRC?

**Comment by ldpgh** [comment]
The TODO comment in `ReferenceNozzleTipConfigurationWizard.jave` is not comming from me.

**Comment by netzmark** [comment]
I didn't use irc of openpnp yet but we can try after the weekend.
On mobile now.

**Comment by ldpgh** [comment]
I would like to prevent to [take over the ownership](https://github.com/openpnp/openpnp/issues/235#issuecomment-431827782)  of this issue, but I can prepare a PR.

Changes will be ...

1. Rotation parameter for calibration
1.1. Convert `angleIncrement` hard-coded into XML-parameter.
1.2. Add  parameters  `angleStart` & `angleStop `as companion to `angleIncrement` as described in the [feature request #784](https://github.com/openpnp/openpnp/issues/784).
1.3. Make those parameters accessible by machine.xml. I need those additional parameters to overwrite the SrcCode settings. My local implementation is limited to an rotation of +/-100deg. 
1.4. Add TODO comment to add angle-parameter fields to the calibration panel.

2. Subtract the Camera-Location from the new location as reported [above](https://github.com/openpnp/openpnp/issues/235#issuecomment-429617753)

3. Workaround of the Division-by-zero as reported [above](https://github.com/openpnp/openpnp/issues/235#issuecomment-429578368). It is not a fix. and will just skip the error condition. Further work is necessary.

4. Keep the pretty verbose debug-messages ... at least at the beginning.

5. Return nozzle/nozzle-tip after calibration to start-position (usually 0 degree).

**Comment by mgrl** [comment]
Made an attempt to fix this in https://github.com/mgrl/openpnp/tree/runoutFix 

The principle is to fit a circle into the runout path by means of kasa algorithm. Should work in theory, will test on the machine in the next hours... It still needs cleanup, optimization and of course verification. Maybe others want to test, too. Or somebody finds obvious mistakes?

**Comment by netzmark** [comment]
I'll be happy to join to the tests since tomorrow.
But could you pls to place here one final debug picture with found and drawn the contour, and place the pipeline you use? Do you search for an outer contour or the inner hole?

**Comment by mgrl** [comment]
currently testing the pipeline - using the built-in pipeline for now.

Am 13.01.2019 um 15:16 schrieb netzmark:

**Comment by netzmark** [comment]
Ok. But still actual question "outer contour or the inner hole of the nozzle"?

**Comment by ldpgh** [comment]
> Ok. But still actual question "outer contour or the inner hole of the nozzle"?
As far as I can tell ... both are working. Just circle center point is used.

**Comment by netzmark** [comment]
Sure they both must be working. I was just curious what others use and find as more effective...

**Comment by ldpgh** [comment]
How about IRC right now ?

**Comment by mgrl** [comment]
so here a capture of the measured points plus the results of the circle fitting by means of the kasa method: 
[runout export.zip](https://github.com/openpnp/openpnp/files/2753308/runout.export.zip) (libreoffice calc)

it seems the circle is well approximated, but if I apply the correction (see updated repo https://github.com/mgrl/openpnp/tree/runoutFix) there will still be offsets - I have too investigate where these may come from...

**Comment by ldpgh** [comment]
MyTestProcedure-1:
- Run the Calibration
- Go to BottomCamera and jog C for (angleStart->angleStop by angleIncrement)

Do we have more simple tests?

**Comment by mgrl** [comment]
so I think I got the math right, but it's not very stable yet... have to find edge cases...

but in general I think I'm on the right way...

no compensation:
![withoutcomp](https://user-images.githubusercontent.com/3868450/51109155-99b8e080-17f4-11e9-85e6-fc0ef7783435.gif)

with compensation:
![withcomp](https://user-images.githubusercontent.com/3868450/51109156-99b8e080-17f4-11e9-8250-f57fec428079.gif)

**Comment by netzmark** [comment]
Wow, congrats, it looks really good :-).
btw, your non compensated movie presents the runout like the best nozzle that I have :-) :-), all the rest are much much worse...
How looks the no-stability problem you say?

**Comment by netzmark** [comment]
So it is not problem of your code but global.
I will be trying with hought circles for fiducials a bit later today.
I think it may be also good to check it for the nozzles, both cases we need to detect good circles. Unfortunately, for nozzles, an outer diameter is changing a lot between 502 nad 506-507. That is what I've asked you what do you detect, outer or inner edge, because inner diameter changes less than outer.

**Comment by mgrl** [comment]
for a 505 I just used this pipeline, that seems to work stable for my machine: 
<cv-pipeline>
   <stages>
      <cv-stage class="org.openpnp.vision.pipeline.stages.ImageCapture" name="9" enabled="true" settle-first="true"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.ConvertColor" name="1" enabled="true" conversion="Bgr2Gray"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.BlurMedian" name="4" enabled="true" kernel-size="5"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.MaskCircle" name="3" enabled="true" diameter="250"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.Threshold" name="2" enabled="false" threshold="15" auto="false" invert="false"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.MaskCircle" name="5" enabled="false" diameter="100"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.SimpleBlobDetector" name="resultsb" enabled="false" threshold-step="5.0" threshold-min="0.0" threshold-max="60.0" repeatability="2" dist-between-blobs="5.0" color="false" color-value="0.0" area="true" area-min="20.0" area-max="5000.0" circularity="false" circularity-min="0.800000011920929" circularity-max="-1.0" inertia="true" inertia-ratio-min="0.5" inertia-ratio-max="1.0" convexity="true" convexity-min="0.4" convexity-max="1.0"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.DetectCirclesHough" name="results" enabled="true" min-distance="50" min-diameter="100" max-diameter="200" dp="1.0" param-1="150.0" param-2="20.0"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.ImageRecall" name="0" enabled="true" image-stage-name="9"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.DrawKeyPoints" name="7" enabled="false" key-points-stage-name="results">
         <color r="255" g="0" b="204" a="255"/>
      </cv-stage>
      <cv-stage class="org.openpnp.vision.pipeline.stages.DrawCircles" name="6" enabled="true" circles-stage-name="results" thickness="1"/>
      <cv-stage class="org.openpnp.vision.pipeline.stages.ImageWriteDebug" name="8" enabled="true" prefix="runoutCalibration" suffix=".png"/>
   </stages>
</cv-pipeline>


and the result: :100: 
with compensation:
![withcomp505](https://user-images.githubusercontent.com/3868450/51113193-87916f00-1801-11e9-83b8-8ca364c95fe5.gif)

without compensation:
![withoutcomp505](https://user-images.githubusercontent.com/3868450/51113194-882a0580-1801-11e9-8a24-457e4dc0d931.gif)

**Comment by netzmark** [comment]
Ech, forgotten we can use specific pipeline to specific nozzle tip...
So now you have used hough circles and seems be stable.
Before not stable was with default SimpleBlobDetector?

**Comment by mgrl** [comment]
I have used simpleBlob before for the small nozzles, but sometimes it didn't find anything. The houghCircle seems to work now properly if tuned for the size of the circle it should detect. Otherwise the circles are jumping and this gives bad results...

I submitted a pull request: #804. Maybe you could test if it works for you in general and then we will see whether we can go on or just disable the feature again until it's really really fixed...

**Comment by netzmark** [comment]
Sure, I'm just going to apply your changes to my machine code and try to test it tomorrow.

**Comment by netzmark** [comment]
Michael, you must have seen this for sure...
https://github.com/openpnp/openpnp/pull/803

I'm going to test both approaches tomorrow and give some feedbacks.

**Comment by mgrl** [comment]
I dont really know what the differences are in #803 since the files seem to be exchanged in total. I guess both approaches won't work together

**Comment by ldpgh** [comment]
It seems space/TAB respective line-ending has been changed by Eclipse.
Turn-off space, TAB, line-endings for the diff ... it should show what you like to see.

**Comment by netzmark** [comment]
No need to have them working together, but test them both to find which version works better. I just have compiled two separated jars, one using your approach and second using @ldph code. Tomorrow I'm going to try to play with them both on my machine.
Another story is Jason's opinion what he finds as better :-).

**Comment by vonnieda** [comment]
@mgrl @ldpgh @netzmark As you all have been working on this I'd really appreciate if you can all come to an agreement and produce one complete PR, or at least if it has to be multiple PRs make sure they don't conflict with each other. I'm not in a position right now to really test any of this, so I need to take someone's word for it that it works :)

**Comment by netzmark** [comment]
To be honest, It's not my job at all, I can only join to the tests. 
These are quite differential codes, I think not possible or sensual to connect them but need to find which works better (and it also will be not very easy) and merge it to the official release. What shall we do if both work well? 😀

Michael, @ldpgh, maybe it's worth to you both do the same that me and try to test both versions...

**Comment by mgrl** [comment]
I just skimmed through #803 (thanks for advice hiding whitespace diffs). It seems @ldpgh added some parameters, etc, but did not change the underlying math? Actually I didn't dive too much into the code/algorithm how it was until today, since it was blamed not to work. so I implemented my own version with the math of which I think how the mechanics work. the animated images proof that the math seems to do it's magic right ;)

I will try the changes from @ldpgh tomorrow, too.

anyway. further tests are needed of course. thats why I send the pull request...

**Comment by ldpgh** [comment]
Can't say it better ... "anyway. further tests are needed of course. thats why I send the pull request..."

@vonnieda ... agree with you.

**Comment by ldpgh** [comment]
Have fixed the issues of my PR (line-endings for diff, variables renamed to pass Travis-CI).
Now the diff-tool can show you the changes easily.
Test on the machine still pending ... too late today.

**Comment by mgrl** [comment]
## without compensation
![without](https://user-images.githubusercontent.com/3868450/51180932-110c7400-18ca-11e9-8518-aff180ec30d5.gif)

## with ldpgh's compensation
![ldpgh](https://user-images.githubusercontent.com/3868450/51180930-110c7400-18ca-11e9-8d0f-617c10c7f9c7.gif)

## with mgrl's compensation
![mgrl](https://user-images.githubusercontent.com/3868450/51181050-5df04a80-18ca-11e9-887b-b25f2942505b.gif)

for reference here is the used pipeline and the debug image-output to see that the pipeline detected in both cases the right center of the nozzle.
[calibration-stage.txt](https://github.com/openpnp/openpnp/files/2759557/calibration-stage.txt)
[runout-ldpgh-rawimages.zip](https://github.com/openpnp/openpnp/files/2759558/runout-ldpgh-rawimages.zip)
[runout-mgrl-rawimages.zip](https://github.com/openpnp/openpnp/files/2759559/runout-mgrl-rawimages.zip)

how did it work for you? I found still a little runout in the compensation method of ldpgh (but better than without).

**Comment by markmaker** [comment]
Hi

@mgrl  / @ldpgh very unfortunate you both invested precious work into this.

Just from understanding the calculation method and commented code better, I slightly favour #804. I hope you understand, @ldpgh. 

I feel this solution is more expandable in the future, as all the fitting is done **after** the complete vision phase in a separate math phase here
https://github.com/openpnp/openpnp/pull/804/files#diff-3903f2851bc6f48dda1b8e8d72c06b40R461
and here
https://github.com/openpnp/openpnp/pull/804/files#diff-3903f2851bc6f48dda1b8e8d72c06b40R511
E.g. if you decide to add ovality later, it can be done without changing/breaking much. 

One suggestion would be to add the number of vision measurements to the UI Panel and then control `angleIncrement` with it. 

I also suggest defaulting this to a prime number (e.g. 7) plus a non-zero start angle, so no aliasing with the video pixels happens. Most cameras have some image sharpening built in, which effectively shifts any edges to the next pixel boundary, which may matter with these small offsets.  So you need to even that out by using angles that do not add up to algebraic angles. 
https://en.wikipedia.org/wiki/Trigonometric_functions#Algebraic_values 

_Mark

**Comment by mgrl** [comment]
I don't see any problem to integrate both versions into one branch one day.
acutally it should not need any other compensation model, since the runout per mechanic must be a circle, must not?

I made some placements on the demoboard, no bottom camera vision:
## without compensation
![compwithout](https://user-images.githubusercontent.com/3868450/51184448-985ee500-18d4-11e9-956f-2d5238496e28.jpg)

## with mgrl compensation
![compwith](https://user-images.githubusercontent.com/3868450/51184449-985ee500-18d4-11e9-8c84-f6e40aa0fddb.jpg)

I guess I need a nozzle with much runout :laughing:

**Comment by netzmark** [comment]
I've tried to perform some tests today but had no too much time for it. However I made some and firstly found that I have serious problem with the light volume to properly detect the nozzle tip. I had to increase exposure seriously to get any result but still get many index0 (no find) answers. Or if I set houghCircles wider I often find the circles with other size I need an moved from the centre - so my process for nothing.
I will try more tomorrow. However I think that it could be usable to finally add some scripts calling or actuators before/after calibration to can control the power of lightning, or auto increasing /decreasing the exposure around the calibration.

**Comment by markmaker** [comment]
> acutally  it should not need any other compensation model, since the runout per mechanic must be a circle, must not?

Within the precision that matters (and/or what vision resolution can get) you're probably right. Theoretically the axis could be tilted, resulting in a slight ovality. But then again vision gets the nozzle tip in one plane but placement on the PCB will be in another plane (due to spring loaded tip) and the ovality would actually work the other way. 

_m

**Comment by mgrl** [comment]
so now I tested to go lower with the number of measurements. it seems the numbers are quite stable up to angleIncrement = 120°. especially the radius is important as it represents the runout. 3 captures are required to fit the circle (which is true for 120°). I can go this low with my implementation, since I fit a circle into the measurements and thus can calculate the exact offset for any possible angle later. The math before did just linear interpolation between two measurements and will be more unaccurate if the the number of measurements is lowered.
additional averaging for my method is not needed since the kasa fit algorithm does its best to fit the circle.

##angle increment 5° (72measurements)
[runoutFix]calculated nozzleEccentricity: centerX: 0.1696830100348604 centerY: 0.22550912581345472 radius: 0.14118548648603493
##angle increment 15° (24measurements)
[runoutFix]calculated nozzleEccentricity: centerX: 0.18089478065314937 centerY: 0.17026865964700014 radius: 0.1401856592249164
##angle increment 60° (6 measurements)
[runoutFix]calculated nozzleEccentricity: centerX: 0.18855007699737564 centerY: 0.1735530454622073 radius: 0.13638914441623307
##angle increment 90° (4 measurements)
[runoutFix]calculated nozzleEccentricity: centerX: 0.16668617902522648 centerY: 0.20960057904441326 radius: 0.13382306133470973
##angle increment 120° (3 measurements)
[runoutFix]calculated nozzleEccentricity: centerX: 0.17139662707281267 centerY: 0.222508790619469 radius: 0.14785716127571277
[runoutFix]calculated nozzleEccentricity: centerX: 0.17649940371794873 centerY: 0.22075060000000002 radius: 0.14390923565418187

**Comment by ldpgh** [comment]
Did a [calibration movie](https://github.com/ldpgh/rasenmaeher/blob/master/openpnp_btmVisionCal.mp4) ...

**Comment by mgrl** [comment]
Looks good. I went through the old code and it seems, the only thing that was wrong is that it centered the nozzle before calibration... That's fixed with your code. Anyway, how to go on? Shall we implement both approaches? We need to define the functions then and just switch between them...

**Comment by ldpgh** [comment]
Our 2 solutions
- table-based + linear interpolation at run-time ... maximum flexibilty, can be tailored at run-time to some degree (eg. lower step size during calibration to reduce interpolation issues)
- model-based ... with model hard-coded in the sources, but less sensitive to measurement errors during calibration. In addition it will not have an overflow issue, if the rotation exceeds the calibration range.

---

With both lists above I really vote to keep both solutions in parallel. Based on its nature the solution are able to address different difficulties.

It is just one more field in  `ReferenceNozzleTipConfigurationWizard.java` to select one or the other approach.

---

When it comes to implementation ...
- `ReferenceNozzle.java`
   - In order to switch between the 2 bottom vision calibration implementations I think those modifications should be in `ReferenceNozzleTip.java`
   - ... if possible
- `ReferenceNozzleTipConfigurationWizard.java`
   - keep my changes. It has your modifications. 
   - I really like to keep the 0degree-offset shown in panel to easily qualify my different nozzle tips.
   - Implemenation relies just on:
      - isCalibrated()
      -  nozzleTip.getCalibration().getCalibratedOffset(rot);
- `ReferenceNozzleTip.java` 
   - @ ldpgh
      - There is just one method added to allow averaging around `findCircle()`.
      - All other stuff should be just "fixes".
      - adding some parameters (`angleStart`, `angleStop`, `angleAverage`) as companions to `angleIncrement`
   - @ mgrl
      -  ... if possible add some notes here in my comment ...
      - Btw.  `angleIncrement` can be changed in machine.xml already.

There is big question mark to me how to handle 2 separate implemenations in parallel for a single API in Java. I'm more a spare time java programmer.

**Comment by mgrl** [comment]
Hi, okay, further we need an documentation in the wiki that gives the users the ability to make a reasonable choice about which algorithm to use and what the parameters/calculated numbers mean. what are limitations, when works which method better... (maybe a chapter here https://github.com/openpnp/openpnp/wiki/Setup-and-Calibration:-Nozzle-Setup or a separate site).


> I really like to keep the 0degree-offset shown in panel to easily qualify my different nozzle tips.

maybe we can rename the field to "runout" or "radius" and you calc the r=sqrt(x2+y2). that field I could use, too.


how I would implement it for now is:
 - remove other result processing than houghcircles from findCircles, but make the processing of houghcircles good, catch errors, etc. in my testing houghcircles worked well, the other method not that good. if there is a need, these functions can be readded later.
- expose angleIncrement to the user in form of number of measurements that shall be taken
- angleStart / angleStop I would leave out, as it seems this is unique to your machine. anyway it remains to be changeable in the machine.xml.
- I would try to isolate our methods into functions that can simply be switched by an if. which method to be taken is dependent on what the user chose.

If you allow, I would try to integrate your changes into my branch and do some cleanup? I don't really know how to do else...

**Comment by ldpgh** [comment]
> I tried to integrate both algorithms but the way I do it with my limited java capabilities makes it quite messy.

If your algoritm is not covering this limitation ... no problem. In this case the user chooses the original table-based implementation.

> Why has your machine only +-90° nozzles? That's the only case, that my algorithm can't handle since the kasa fit works better for full circles...

It is an resolution limit.
With `pick @ -90deg and place @ +90deg` and `pick @ +90deg and place @ -90deg` the whole 360deg are covered. In reality the limit is +/-100deg to have margins for misaligned parts in the SMD-strip.

**Comment by mgrl** [comment]
okay, I'm on the way... it's getting complexer and complexer 😞  

I got both approaches integrated into one now. 

First makeup for the GUI is done, too:

![unbenannt](https://user-images.githubusercontent.com/3868450/51351255-a6fef500-1aaa-11e9-848c-c83a5afaace9.PNG)
(switching doesn't work yet, but under the hood both algorithms coexist and should be usable).  will continue soon, but have to come to an end

**Comment by mgrl** [comment]
ignore me, these are for wiki documentation
![pipeline3](https://user-images.githubusercontent.com/3868450/51399984-8c2e8e00-1b47-11e9-92df-b6ac2bddb79b.PNG)
![pipeline1](https://user-images.githubusercontent.com/3868450/51399985-8c2e8e00-1b47-11e9-9cf0-c20e6b3cf8ad.PNG)
![pipeline2](https://user-images.githubusercontent.com/3868450/51399987-8c2e8e00-1b47-11e9-8c37-0f9d9148300d.PNG)

**Comment by mgrl** [comment]
and here the current gui:

![unbenannt](https://user-images.githubusercontent.com/3868450/51401080-3ad3ce00-1b4a-11e9-8e42-7d062c46929a.PNG)

**Comment by mgrl** [comment]
[this is for the wiki](https://github.com/openpnp/openpnp/wiki/Runout-Compensation-Setup)
![algorithm-graphic](https://user-images.githubusercontent.com/3868450/51412842-3b319080-1b6d-11e9-9642-f20b5d05f8a4.png)

**Comment by mgrl** [comment]
ah, and what we can see if we compare the first 3 images with the second 3:
the centerX/Y on my machine is not yet exactly at 0/0, at ldpgh it is at 0/0. So I have either

1. the bottom camera position not 100% right or
2. the nozzle offset not 100% right.

this information is delivered only from the modelbased algorithm and can help users to eliminate position offset problems. but we can't reliably find out, whether its 1. or 2. unfortunately

**Comment by ldpgh** [comment]
> ah, and what we can see if we compare the first 3 images with the second 3:
> the centerX/Y on my machine is not yet exactly at 0/0, at ldpgh it is at 0/0. So I have either
> 
> 1. the bottom camera position not 100% right or
> 2. the nozzle offset not 100% right.

Though already closed/Just for completeness  ... IMHO the items 1. respective 2.  describe an open issue.

At the moment I define/use the very first nozzle-tip as reference nozzle-tip having zero offset by definition. In this case it covers the nozzle-camera offset as well and the bottom vision camera is adjusted by the so-called reference nozzle-tip.

But is paid by the tool changer. Any nozzle-camera offset makes it hard to capture any nozzle-tip sitting in the tool changer by the down-lookong camera ... except your offset is pretty small (<< 0.1mm).

At the moment I manually modify the captured tool changer locations.

**Comment by aneox** [comment]
Nice feature! thanks!

But why now head moves to bottom camera in to step?
Is second microstep needed? why not add compensation in first step?
I didnt look in code, but found this behavior in my machine now

**Comment by mgrl** [comment]
Have you calibration enabled? You would actively need to enable it to see a difference. If you enabled it, it calibrated right after fiducial check. 

I will add this to the wiki information page...
https://github.com/openpnp/openpnp/wiki/Nozzle-Tip-Runout-Compensation-Setup

**Comment by aneox** [comment]
@mgrl 
Sorry, i forgot to say, calibration works well, i told about part alignment in job proccess.
After calibration enabled, it start to work in 2 move steps. First head moves to bottom camera location, second some compensation move, then part alignment starts.

**Comment by mgrl** [comment]
Hmm, could you upload some logs? I don't think the second move is due to this feature... Runout compensation just modifies the destination location and never invokes a movement on it's own in normal operation...

**Comment by aneox** [comment]
@mgrl 
I have merged only 2 features, affine fiducial transform and this.
Didnt find anything interesting in log, only additional moveTo from nozzle.
Do you use option "Rotate parts prior to vision?" in bottom vision settings? I do.

**Comment by vonnieda** [comment]
I do not recommend merging affine transforms at this time. It had
fundamental flaws that can cause serious errors. Primarily, it does not
respect units with regard to transforming locations.

**Comment by aneox** [comment]
@vonnieda 
Thank you! I will use it with attention)
I have found some time to test and first board show good result with it.

**Comment by vonnieda** [comment]
I'm actually working on that code *right now* and hope to release it into
the main branch in next few days.

Jason

**Comment by netzmark** [comment]
Hi Jason, "affine transforms" do you plan to release in next few days, did i understand well?? It will be another great improvement!

**Comment by vonnieda** [comment]
@netzmark Yes, that's correct. First version in develop branch by end of day, I think.

**Comment by netzmark** [comment]
Excellent, I'll merge it to my custom and start the tests at once.

**Comment by netzmark** [comment]
Excellent, I'll merge it to my custom and start the tests at once.

**Comment by aneox** [comment]
Hi! In my opinion this approach is hard and overhead. May i offer my variant.

When job starts or nozzle change, we calibrate nozzle offset with bottom vision at 0 rotation.
When job runs, we pick all parts with 0 rotation and rotate it after pick, at safe Z.
When bottom vision check part align, we do not see nozzle, so nozzle runout is not interesting. 
Vision find part offset and place correctly.

**Comment by mgrl** [comment]
Am 20.03.2019 um 09:56 schrieb aneox:
>
> Hi! In my opinion this approach is hard and overhead. May i offer my 
> variant.
>
> When job starts or nozzle change, we calibrate nozzle offset with 
> bottom vision at 0 rotation.
> When job runs, we pick all parts with 0 rotation and rotate it after 
> pick, at safe Z.
> When bottom vision check part align, we do not see nozzle, so nozzle 
> runout is not interesting.
> Vision find part offset and place correctly.
>
The algorithm is intended to replace bottom vision by a one time nozzle 
calibration which is sufficient for most of the parts (resistors, etc). 
If you use bottom vision on all parts, you could just ignore the feature 
and calibrate your machine to properly pick up all parts. The rest is 
left to the bottom vision then.

**Comment by netzmark** [comment]
What?? Do you suggest to to nozzle calibration and then do pnp without bottom vision?

**Comment by aneox** [comment]
@mgrl i am using your autofeeders, thank you for share it!
And i see that part pick location not perfect, parts in tape has little offset and sometimes feed not perfect. So parts often picks with not perfect center align and iam sure, bottom vision is must have if you need perfect placement, especially for 0402.

**Comment by mgrl** [comment]
For the bigger passives, why not? For sure not with 0201, and maybe not 
with 0402. Minimum size I use is 0603 and with these I had sufficient 
placement accuracy...

Am 20.03.2019 um 10:12 schrieb netzmark:

**Comment by netzmark** [comment]
Hmmm, maybe you're right it's good for middle size parts as you say. Not for 0402 and not for fine pitch chips.

**Comment by aneox** [comment]
For now, without your calibration, after nozzle change and if nozzle has big runout, pick of 0402 not stable, because of not stable nozzle offset. If pick always with 0 rotation and calibrate nozzle offset with 0 rotation each nozzle change, pick of small parts will be more stable.

With my camera, your nozzle calibration not stable and always result with different runout. 
If calibrate only one point, nozzle offset in 0 rotation, it will be much more easier and stable i think.

**Comment by mgrl** [comment]
Am 20.03.2019 um 10:34 schrieb aneox:
>
> With my camera, your nozzle calibration not stable and always result 
> with different runout.
>
Whats not stable? The pipeline? It works best if the pipeline returns 
only exactly one result instead of a list. I would need logs and debug 
images to see whats going on. As a fallback you still have the old 
implementation (table based) if it's not the pipeline.
>
> If calibrate only one point, nozzle offset in 0 rotation, it will be 
> much more easier and stable i think.
>
Openpnp is not designed to always pick up at 0 degree - there must be a 
reason for that, must not? Maybe Jason knows.

**Comment by aneox** [comment]
Yes, pipeline is not stable. Because my camera has big focal lens to work with BGA chips, so nozzle tip is very small in image, hard to make it stable. Maybe i need second camera for BGA, not sure if it supports for now. But current camera is enough to works with 0402, because parts has shiny surface.

Openpnp may be redesigned to make it better)) 
Rotate parts after pick at safe Z doesnt have disadvantages i think.

**Comment by aneox** [comment]
I think if your machine cannot place 0603 without bottom vision and visual calibration, upgrade nozzle holder or nozzle or etc.. its better way, because price for it not high. I think visual calibration needs for small fine pitch parts )

**Comment by mgrl** [comment]
Am 20.03.2019 um 10:56 schrieb aneox:
>
> I think if your machine cannot place 0603 without bottom vision and 
> visual calibration, upgrade nozzle holder or nozzle or etc.. its 
> better way, because price for it not high. I think visual calibration 
> needs for small fine pitch parts )
>
Oh, it can for sure... But more accuracy doesn't hurt if its for free - 
and if it's just for picking up parts...

**Comment by aneox** [comment]
i agree, of course, may be you understand me wrong, so for my english.

**Comment by netzmark** [comment]
I have camera covering 0402 and BGA 30x30mm.
No problem with stable pipeline if you always bring centre of physical axis to the bottom camera (then it doesn't add xy offset centre physical to centre found) and you spend some time on pipeline tuning. You must narrow size range of expected nozzle in pipeline and tune the rest of pipeline to each time find proper diameter. If your tip is ie 0.8mm you won't get good effect setting range from 0.5-1.3 instead og optimalized the rest of pipeline. Because pipeline will "find" some diameter at each probe but it will be not proper diameter and it will be drawn with center moved xy  from real centre. If your tip is 0.8, set range 0.9-1.0 and play to be found 100/100. Then you'll get repetitions +/-0.02mm what is more than need for the good picking 0402.

**Comment by aneox** [comment]
I understand that pipeline tuning is important, i agree, thank you for advise. I will try tune it better.
My offer was to reduce amount of camera shots from 6 to only one, to reduce possible error.
